solr/core/src/java/org/apache/solr/cluster/events/impl/DelegatingClusterEventProducer.java

```diff
@@ -90,7 +101,11 @@ public void setDelegate(ClusterEventProducer newDelegate) {
         log.debug("--- delegate {} already in state {}", delegate, delegate.getState());
       }
     }
-    this.version++;
+    Phaser localPhaser = phaser; // volatile read
+    if (localPhaser != null) {
+      assert localPhaser.getRegisteredParties() == 1;
+      localPhaser.arrive(); // we should be the only ones registered, so this will advance phase each time
```

Comment:
Why make the assumption this class is the only one used by the test? Maybe the test uses the `Phaser` to synchronize multiple classes? That just means remove the assert and the comment.

But here's the issue I alluded to in the Jira: by exposing directly `Phaser` and not a test provided implementation of an interface method to call, we restrict the usage of such synchronization to only allow the test code to know when something took place in the production code.
If we expose an interface, it allows the test to stall the production code at this stage, maybe to make it wait for another condition to happen (for example let the test verify the current state produced by the production code) and then the test can decide to resume the production code. This is useful when verifying state machines for example, as otherwise the state transitions can be so fast that it's impossible to verify reliably (by the time the test is notified of the first transition and tries to verify it, the state machine has made 5 more transitions...).

I'd really prefer to expose an interface as in the previous version of the code. The have the test implement it by basically delegating to `Phaser`.

Reply:
While I agree that all of those things are useful sounding in theory, we don't ostensibly need them right now, and I'd prefer to keep the changes smaller. I don't like the idea of adding an interface with a single test implementation on the hunch that someday somebody else might want to do something else with it. If that's the case, we can refactor it in the future.

Reply:
In the JIRA issue you specifically suggested "maybe the test can provide a latch" which would be a specific synchronization implementation, but a latch happened to be insufficient because we need something that can block for us multiple times. I don't see this as too divergent from that idea, and have seen other code available for injecting latches specifically. It feels the same.

Reply:
Fair enough.
Looks good to me!